src/number.c

```diff
@@ -417,3 +417,91 @@ scm_p_remainder(ScmObj _n1, ScmObj _n2)
 
     return MAKE_INT(n1 % n2);
 }
+
+SCM_EXPORT
+ScmObj scm_p_gcd(ScmObj left, ScmObj right, enum ScmReductionState *state)
+{
+    scm_int_t result, n1, n2, tmp;
+    DECLARE_FUNCTION("gcd", reduction_operator);
+
+    result = 0;
+    switch (*state) {
+    case SCM_REDUCE_PARTWAY:
+    case SCM_REDUCE_LAST:
+        ENSURE_INT(left);
+        result = SCM_INT_VALUE(left);
+        /* Fall through. */
+    case SCM_REDUCE_1:
+        ENSURE_INT(right);
+        /* BEGIN code for computing gcd of two numbers
+	   result = GCD(result, right); */
+
+	n1 = result;
+	n2 = SCM_INT_VALUE(right);
+	
+	if (n1 < 0) n1 = - n1;
+	if (n2 < 0) n2 = - n2;
+	
+	while (n1 != 0) {
+	    tmp = n1;
+	    n1 = n2 % n1;
+	    n2 = tmp;
+	}
+	result = n2;
+	/* END code for computing gcd of two numbers */
+	
+        /* Fall through. */
+    case SCM_REDUCE_0:
+        break;
+    default:
+        SCM_NOTREACHED;
+    }
+
+    return MAKE_INT(result);
+}
+
+SCM_EXPORT ScmObj
+scm_p_lcm(ScmObj left, ScmObj right, enum ScmReductionState *state)
+{
+    scm_int_t result, big, small, n1, n2;
+    DECLARE_FUNCTION("lcm", reduction_operator);
+
+    result = 1;
+    switch (*state) {
+    case SCM_REDUCE_PARTWAY:
+    case SCM_REDUCE_LAST:
+        ENSURE_INT(left);
+        result = SCM_INT_VALUE(left);
+        /* Fall through. */
+    case SCM_REDUCE_1:
+        ENSURE_INT(right);
+	
+	/* BEGIN code for computing gcd of two numbers
+	   result = LCM(result, right); */
+
+	n1 = result;
+	n2 = SCM_INT_VALUE(right);
+	
+	if (n1 < 0) n1 = - n1;
+	if (n2 < 0) n2 = - n2;
+	
+	big    = n1 > n2 ? n1 : n2;
+	small  = n1 < n2 ? n1 : n2;
+	n1 = big;
+    
+	while (n1 % small != 0)
+	    n1 += big;
+
+	result = n1;
```

Comment:
How about defining `static scm_int_t gcd(scm_int_t x, scm_int_t y)` and using it in `scm_p_gcd()` and `scm_p_lcm()`?